balancer_wrapper.go

```diff
@@ -374,3 +411,42 @@ func (acbw *acBalancerWrapper) closeProducers() {
 		delete(acbw.producers, pb)
 	}
 }
+
+// RegisterHealthListener stores a health listener from the LB policy. It sends
+// updates to the health listener as long as the SubConn's connectivity state
+// doesn't change and a new health listener is not registered. To invalidate a
+// the currently registered health listener, acbw updates the connectivity state
+// stored by the object. If a nil listener is registered, the active health
+// listener is dropped.
```

Comment:
> The health listener provided by the LB policy is not really stored anywhere.

Updated the comment to say "accepts a health listener" . Technically speaking, the health listener is captured by the closure that is scheduled on the serializer. So we can say it's stored, but the scope is restricted to the closure 🙃 .

> There is only one health update ever sent to the listener.

The health service client integration will come in a followup PR. That will send multiple health updates.

> This is only true if the previously registered health listener was not yet invoked.

It is possible that a health listener is registered after the health update callback checks the healhData but before it calls the health listener. I didn't keep the healthMu lock to avoid a deadlock if an LB policy tries to register a health listener synchronously while handling a health update. However, this scenario doesn't seem useful, so I changed the code. Now healthMu is held while calling the health listener. This means that after `RegisterHealthListener` returns, it is guaranteed that the old listener will not be called again. I've updated the balancer.RegisterHealthListener comment to mention the following
```go
// This method must not be called synchronously while handling an update
// from a previously registered health listener.
```

The 3 points you mentioned at the end are the implementation details to provide the guarantees mentioned in the  SubConn interface's method. Things are done this way so that the sequence of events delivered to the LB Policy are consistent with the way the LB policy sees the SubConn connectivity state. For e.g, the SubConn state transitions as follows:
1. CONNECTING // Delivered to LB policy
2. READY // Delivered to LB policy
3. IDLE // Queued

The LB policy has seen only 1 & 2. LB policy registers a health listener. The ready update on the listener is queued . The event sequence becomes:
1. CONNECTING // Delivered
2. READY // Delivered
3. IDLE // Queued
4. Health READY // Queued

Now the connectivity update for IDLE is delivered, this needs to invalidate the upcoming health update. Otherwise, the LB policy will see a health update for READY for a subchannel it sees as IDLE.

1. CONNECTING // Delivered
2. READY // Delivered
3. IDLE // Delivered
4. Health READY // Skipped

I feel that specifying this level of detail comments is not too useful for readers who just want to know what the method does. The code is the final source of truth, so readers can always read the code to know the implementation details.